### PR TITLE
Use a fresh tf.Graph for each data augmentation batch

### LINKED ISSUE #2726: TF image augmenter leaks memory
- OD in master crashes due to a memory leak

### PR DESCRIPTION
**Pull request by nickjong**
Fixes #2726

### CODE REVIEW

File: src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py [hunk 1]
```diff
@@ -22,20 +22,27 @@
...
-    else:
+
+def get_augmented_data(images, annotations, output_height, output_width):
+
+    output_shape = (output_height, output_width)
+
+    # Suppresses verbosity to only errors
+    tf.logging.set_verbosity(tf.logging.ERROR)
+
+    with tf.Session() as session:
```
Comment (Jarvi-Izana) [inline src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py]: Could you have any explanation of this? It seems very rare that TF won't recycle unused sessions.

Comment (shreyajain17) [reply src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py]: On digging a bit, I have found resources that state that session leaks may occur if sessions are not closed properly. I am trying to understand a bit more by reproducing it and looking at the memory footprint.

File: src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py [hunk 2]
```diff
@@ -27,7 +27,14 @@ def get_augmented_data(images, annotations, output_height, output_width, resize_
     # Suppresses verbosity to only errors
     tf.logging.set_verbosity(tf.logging.ERROR)
 
-    session = tf.Session()
+    graph = tf.Graph()
+    with graph.as_default:
+        with tf.Session() as session:
```
Comment (shreyajain17) [inline src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py]: I just saw this. Wouldn't you have to mention `with tf.Session(graph=graph) as session:`

**Comment by nickjong** [comment]
It turns out the underlying problem here is the same as the sound classifier issues: we keep reusing the default graph, which grows without bound as we add more and more operations. (Unlike the core NN models, we have to extend new graph nodes for each batch, since the precise operations being performed are randomly determined each time.)

The solution is to apply the `Graph` and `Session` types more correctly. See the introductory comments at https://www.tensorflow.org/api_docs/python/tf/Graph and https://www.tensorflow.org/api_docs/python/tf/compat/v1/Session

File: src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py (hunk 2 above)
Comment (nickjong) [reply src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py]: I believe the `graph.as_default()` takes care of that: any attempt to use the default graph, including by the `tf.Session` initializer, will use our local `graph` instance instead

Comment (shreyajain17) [reply src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py]: I think it should be `graph.as_default()`

**Comment by nickjong** [comment]
Ugh, my "cleaned up" version doesn't actually work. More functional patch forthcoming

File: src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py (hunk 2 above)
Comment (nickjong) [reply src/python/turicreate/toolkits/object_detector/_tf_image_augmenter.py]: Yup, I got it working like that in some messed up code and then thought I could clean it up without testing before I pushed, sigh

**Pull request merged**